[Wong Chee Hong] iP

cheehongw opened the pull request:
# DUKE_HEADER
> Wow much quote
* bullet points make me stand out
* point number 2
* ok point taken

imma throw in a divider 💯 
---------------------------------

1. how bout numbered list?
2. ok lets stop at two

```java
class Main {
  if (isAlive) {
      csStudent.code();
  }
}
```

- [x] your favourite task list is back
- [ ] please stop this 😭 
- [X] help 

[please follow my github here tysm](https://github.com/cheehongw)

My CS2030S TA said to me that:
> your `ArrivalEvent` should **inherit** the `Event` class!!!
> i dont see encapsulation!!!

_why is this not working??_

Commit:
Add Gradle support

cheehongw committed:
Ignore java .class files

cheehongw committed:
Add greet, echo and exit

as per Duke level 1 requirements.

cheehongw committed:
Implement `list` and `add`

as per Duke Level 2 specifications.

cheehongw committed:
Add Task

cheehongw committed:
Add `mark`, `unmark`

Add ability to mark and unmark tasks as done, as per Duke level 3
requirements

cheehongw committed:
Split `Task` into `ToDo`, `Event` and `Deadline`

Tasks are of 3 specific types as mentioned in subject header.
Add functionality in Duke to handle the addition of these 3 types of
tasks, as per level 4 requirements.

cheehongw committed:
Add automated text UI testing

as per `A-TextUiTesting` requirements

cheehongw committed:
Add error handling for Duke

cheehongw committed:
Implement `delete`

cheehongw committed:
Add ability to save tasks to hard disk

Currenly, Tasks are not preserved across sessions of Duke.

To preserve state across sessions of Duke, the taskList has to be
written to memory.

Now, any commands that modify the list of tasks will be written to the hard
disk.

cheehongw committed:
Use Date/Time instead of String for dates

Currently, the dates for `Deadline` and `Event` are stored as strings.

We should aim to represent this as a DateTime class instead of a String.

This change stores dates and time as `java.time.LocalDate` and `java.time.LocalTime` objects respectively.

cheehongw committed:
Update .gitignore

cheehongw committed:
Merge branch 'branch-Level-7'

cheehongw committed:
Merge branch 'branch-Level-8'

cheehongw committed:
Merge Level 7 and Level 8 features

After merging level-7 and level-8 to master, the features were
incompatible with each other.

This commit ensures these two features are compatible with each other.

cheehongw committed:
Add Ui class to deal with interaction with user.

cheehongw committed:
Refactor list of tasks into separate class

Previously, an `ArrayList<Task>` was used to contain the list of tasks.
Methods were then declared to manipulate the items in `ArrayList`

However, this is not good practice of encapsulation.

Thus, behaviour involving manipulating the list of tasks and
the `ArrayList<Task>` instance have been refactored into their separate
class.

cheehongw committed:
Rename DukeEngine.java to Parser.java

To meet A-MoreOOP

cheehongw committed:
Put classes into package

Two packages have been created: `duke` and `duke.task`. The latter
consists of `Task.java` and its subclasses. The former contains all
other `.java` file.

cheehongw committed:
Add junit tests

cheehongw committed:
Fix .gitignore weirdness

somehow `/data/` is still being tracked even though it is added into
`.gitignore`

hopefully this will fix it

cheehongw committed:
Fix directory not generating.

Previously, the `data/` directory was not being autogenerated if it doesn't exist on
the system.

This fixes that by calling `mkdir` if the directory doesn't exist

cheehongw committed:
Fix runtest.bat to work with packages

cheehongw committed:
Add JavaDocs

cheehongw committed:
Enforce Coding Standards.

Ensure source code is compliant with coding standards.

cheehongw committed:
Implement `Find`

User can now search for tasks matching a provided keyword.

cheehongw committed:
Merge branch 'branch-Level-9'

cheehongw committed:
Merge branch 'branch-A-JavaDoc'

cheehongw committed:
Merge branch 'branch-A-CodingStandard'

cheehongw committed:
Merge branch 'add-gradle-support'

myc37 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,17 @@
+package duke;
+import java.util.Scanner;
+
```
Could add some Javadoc comments here

myc37 commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,7 @@
+package duke;
+
```
Could add some Javadoc comments here

myc37 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,159 @@
+                Storage.updateTaskFile(tasks);
+                break;
+
+            case "todo":
```
These fallthrough cases aren't necessary

myc37 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,159 @@
+            taskArgs = commandDetails.split(" /at ", 2);
+        }
+
+        if (taskArgs.length < 2) throw new DukeException(String.format(
```
If clause should have curly brackets

myc37 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,159 @@
+    }
+
+    private void assertNonEmptyDetails(String details) throws DukeException {
+        if (details == null) throw new DukeException("Missing details!");
```
If clause should have curly brackets

myc37 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,159 @@
+    }
+
+    private void assertValidItemNumber(String str) throws DukeException {
+        if (str == null) throw new DukeException("Missing item number!");
```
All these if clauses should have curly brackets

myc37 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,159 @@
+        try {
+            tasks = Storage.readSaveFile();
+        } catch (DukeException e) {
+            Ui.printMessage(Ui.CORRUPTED_SAVE_MESSAGE);
```
Good job using named constants 😄

myc37 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,106 @@
+ */
+public class Storage {
+
+    private static final String FILEPATH = "data/duke.txt";
```
Naming convention here should be consistent, either FILEPATH and DIRPATH or FILE_PATH and DIR_Path.

myc37 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,106 @@
+import duke.task.Event;
+import duke.task.Task;
+import duke.task.ToDo;
+
```
Standard java package imports should be grouped together

myc37 reviewed (commented):
Looks good! Just be wary of your if-clause style 😄

cheehongw committed:
Configure `build.gradle` to run Duke

sophiien commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,64 @@
+    
+    public static final String LINE = "----------------------------------------\n";
+    public static final String GREETING_MESSAGE = "Wow! Hello! I'm Duke.\nWhat can I do for you?";
+    public static final String UNKNOWN_COMMAND_MESSAGE = "OoPs! I don't know what that means :P";
```
Good job with the sticking to naming conventions for constants

sophiien commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,99 @@
+
+        try {
+            deadline = LocalDate.parse(datetime[0], DATE_FORMATTER);
+            time = datetime.length == 1 ? null : LocalTime.parse(datetime[1], TIME_FORMATTER);
```
Can consider splitting this rather long ternary expressions into 2 lines

cheehongw committed:
Add checkstyle

sophiien commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,99 @@
+
+        String date = this.deadline.format(dateFormatter);
+        String time = this.time == null ? "" : " " + this.time.format(timeFormatter);
+        return String.format("[D]%s %s (by: %s%s)", this.statusString(), this.task, date, time);
```
Can consider splitting this relatively long statement into more lines

sophiien reviewed (commented):
Very neat and clean code with a lot of OOP. Good job! :)

cheehongw committed:
Update unit tests to be compliant with checkstyle

cheehongw committed:
Merge branch 'branch-A-Gradle'

cheehongw committed:
Merge branch 'branch-A-CheckStyle'

cheehongw committed:
Fix style in files

cheehongw committed:
Level-10 JavaFX tutorial part 1

cheehongw committed:
level-10 tutorial part 3

cheehongw committed:
JavaFX tut 4

cheehongw committed:
Exercise 1 of JavaFX tut 4

cheehongw committed:
Add GUI to Duke

cheehongw committed:
Merge branch 'branch-Level-10'

cheehongw committed:
Refactor parsing of object instances from savefile

Currently, object instances stored in the savefile are parsed and
converted within Storage.

This behaviour should be encapsulated by their respective objects, since
each type has their own unique representation in the savefile.
Furthermore, this behaviour is the inverse of `toFileFormat()`, and
should be placed together with said method.

Let's,
* extract and encapsulate code relevant to converting objects from
their savefile representation underneath their own class in a method
called `fromFileFormat`.
* create a CorruptedSaveException to indicate errors with converting
  from the savefile.

Currently, this change affects `Deadline`, `Event` and `Todo` objects.

cheehongw committed:
Group index-based commands in inputHandler

Currently, index based commands from the user such as `mark`,`unmark`
and `delete` have duplicated code within inputHandler. For example, all
index-based commands will need to check for a valid TaskList index
before proceeding with the operation specified by the user command.

Grouping index-based commands will help improve code quality.

Lets,
* extract relevant to index-based commands into a separate method
`indexNumberOperation` to handle index-based commands. Doing so will
make addition of future index-based commands easier as well.
* throw an `InvalidItemNumberException` instead of `DukeException` to
  represent the exceptional case where the user supplies an invalid item
  number.

With the addition of `InvalidItemNumberException` and
`CorruptedSaveException` from previous commits, there will be a new
package `duke.exceptions` to contain all Duke-related exceptions.

cheehongw committed:
Add output Queue to buffer Duke's output messages

Instead of having a call-response model when a user inputs a command,
use a queue to buffer the output messages by Duke. The application will
then print all messages in the queue until it is empty.

This will allow a command to generate multiple distinct output messages.

cheehongw committed:
Rename package`duke.task` to `duke.tasks`

cheehongw committed:
Make JavaFX work with outputQueue

cheehongw committed:
Add assertions

cheehongw committed:
Merge pull request #3 from cheehongw/branch-A-CodeQuality

Improve Code Quality (A-CodeQuality)

cheehongw committed:
Merge branch 'master' into branch-A-Assertions

cheehongw committed:
Merge pull request #2 from cheehongw/branch-A-Assertions

Add assertions (A-Assertions)

cheehongw committed:
Set up github workflows

cheehongw committed:
Add equals method for each type of Task

cheehongw committed:
Add detection of duplicates (C-DetectDuplicates)

Duke now blocks the addition of a duplicated task.

cheehongw committed:
Merge branch 'branch-C-DetectDuplicates'

cheehongw committed:
Improve code quality.

Refactor magic strings in TaskList.java into named constants.

cheehongw committed:
Enforce checkstyle

cheehongw committed:
Improve GUI, move GUI files into subpackage

Added a title to the Java GUI window, added a box around messages.
Tried to put profile pictures into circles but not very successful.

cheehongw committed:
Add userguide

cheehongw committed:
Merge branch 'A-BetterGui'

cheehongw committed:
Fix storage bug causing tasks to not be saved.

Event and Deadline with both date and time are not stored correctly,
resulting in a parse error when reading from the savefile.

Changes were made to `toFileFormat` of the 2 classes to fix this bug.

cheehongw committed:
Fix dialogbox resizing bug

Long labels in dialogbox gets resized when the scrollPane activates.

See discussion here:
https://github.com/nus-cs2103-AY2122S2/forum/issues/145

cheehongw committed:
Update readme to include pictures

cheehongw committed:
Enforce checkstyle

cheehongw committed:
Merge branch 'branch-A-UserGuide'

cheehongw committed:
Update docs to include release link

cheehongw committed:
Update README.md

Fix table formatting issue with github pages

cheehongw committed:
Delete junit-platform-console-standalone-1.8.2.jar

Remove binaries from github

cheehongw committed:
Fix issue #4

Close #4.
